File: ml-tools/url_features.py
from urllib.parse import urlparse
import re, math

def shannon_entropy(s: str) -> float:
    if not s:
        return 0.0
    probs = [s.count(c) / len(s) for c in set(s)]
    return -sum(p * math.log2(p) for p in probs)


def extract_url_features(url: str) -> dict:
    parsed = urlparse(url)

    hostname = parsed.hostname or ""
    path = parsed.path or ""
    query = parsed.query or ""

    features = {}

    features["url_length"] = len(url)
    features["hostname_length"] = len(hostname)
    features["path_length"] = len(path)
    features["query_length"] = len(query)

    features["num_dots"] = url.count(".")
    features["num_hyphens"] = url.count("-")
    features["num_at"] = url.count("@")
    features["num_slashes"] = url.count("/")
    features["num_digits"] = sum(c.isdigit() for c in url)
    features["digit_ratio"] = features["num_digits"] / max(len(url), 1)

    # Suspicious patterns
    features["has_ip"] = int(bool(re.search(r"\d+\.\d+\.\d+\.\d+", hostname)))
    features["has_https"] = int(parsed.scheme == "https")

    features["hostname_entropy"] = shannon_entropy(hostname)

    return features

def featurize_urls(urls):
    return [extract_url_features(u) for u in urls]

File: src/ml/urlFeatures.js

export function shannonEntropy(str) {
  if (!str || str.length === 0) return 0;
  const freq = {};
  for (const c of str) freq[c] = (freq[c] || 0) + 1;
  const len = str.length;
  let entropy = 0;
  for (const c in freq) {
    const p = freq[c] / len;
    entropy -= p * Math.log2(p);
  }
  return entropy;
}

export function extractUrlFeatures(url) {
  let parsed;
  try {
    parsed = new URL(url);
  } catch {
    return null;
  }

  const hostname = parsed.hostname || "";
  const query = parsed.search ? parsed.search.slice(1) : "";

  return {
    url_length: url.length,
    hostname_length: hostname.length,
    query_length: query.length,

    num_dots: (url.match(/\./g) || []).length,
    num_hyphens: (url.match(/-/g) || []).length,
    num_at: (url.match(/@/g) || []).length,
    num_slashes: (url.match(/\//g) || []).length,
    num_digits: (url.match(/\d/g) || []).length,
    digit_ratio: url.length ? (url.match(/\d/g) || []).length / url.length : 0,

    has_ip: /\d+\.\d+\.\d+\.\d+/.test(hostname) ? 1 : 0,
    has_https: parsed.protocol === "https:" ? 1 : 0,
    hostname_entropy: shannonEntropy(hostname)
  };
}

